spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/embedded/EmbeddedCouchbaseAutoConfiguration.java

```diff
@@ -0,0 +1,151 @@
+import org.springframework.boot.autoconfigure.couchbase.CouchbaseProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.env.MapPropertySource;
+import org.springframework.core.env.MutablePropertySources;
+import org.springframework.core.env.PropertySource;
+import org.springframework.util.StringUtils;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Embedded Couchbase.
+ *
+ * @author Alex Derkach
+ * @since 2.0.0
+ */
+@Configuration
+@EnableConfigurationProperties({CouchbaseProperties.class})
+@AutoConfigureBefore(CouchbaseAutoConfiguration.class)
+@ConditionalOnClass({CouchbaseBucket.class, CouchbaseMock.class})
+public class EmbeddedCouchbaseAutoConfiguration {
+
+	private final CouchbaseProperties properties;
+	private final ApplicationContext applicationContext;
+
+	public EmbeddedCouchbaseAutoConfiguration(CouchbaseProperties properties,
+			ApplicationContext applicationContext) {
+		this.properties = properties;
+		this.applicationContext = applicationContext;
+	}
+
+	@Bean(destroyMethod = "stop")
+	@ConditionalOnMissingBean
+	public CouchbaseMock couchbaseMock() throws IOException, InterruptedException {
+
+		CouchbaseProperties.Bucket bucketProperties = this.properties.getBucket();
+
+		BucketConfiguration bucketConfiguration = new BucketConfiguration();
+		bucketConfiguration.name = bucketProperties.getName();
+		bucketConfiguration.password = bucketProperties.getPassword();
+		bucketConfiguration.type = Bucket.BucketType.COUCHBASE;
+		bucketConfiguration.numReplicas = 1;
+		bucketConfiguration.numNodes = 1;
+
+		CouchbaseMock couchbaseMock = new CouchbaseMock(0, Collections.singletonList(bucketConfiguration));
+		couchbaseMock.start();
+		couchbaseMock.waitForStartup();
+
+		String host = couchbaseMock.getHttpHost();
+		int bootstrapHttpDirectPort = couchbaseMock.getHttpPort();
+		int bootstrapCarrierDirectPort = couchbaseMock.getBuckets().get(this.properties.getBucket().getName()).getServers()[0].getPort();
+
+		couchbaseMock.startHarakiriMonitor(host + ":" + bootstrapHttpDirectPort, false);
+
+		this.properties.setBootstrapHosts(Collections.singletonList(host));
+		this.properties.setBootstrapHttpDirectPort(bootstrapHttpDirectPort);
+		this.properties.setBootstrapCarrierDirectPort(bootstrapCarrierDirectPort);
+
+		setEnvironmentProperties(this.applicationContext, host, bootstrapHttpDirectPort, bootstrapCarrierDirectPort);
+
+		return couchbaseMock;
+	}
+
+	private void setEnvironmentProperties(ApplicationContext currentContext,
+			String host,
+			int bootstrapHttpDirectPort,
+			int bootstrapCarrierDirectPort) {
+		if (currentContext instanceof ConfigurableApplicationContext) {
```

Comment:
This is fragile and should be avoided. It looks like that this auto-configuration should expose a `CouchbaseConfiguration` rather than augmenting the Environment so late in the game.

If we opened `CouchbaseConfiguration` a bit (make that its own class with the necessary parameters that you need), then we could expose one here and leave the original auto-configuration unchanged (it would simply backoff as it does already prior to this change).

Reply:
fixed